### PR TITLE
Adds an optional context types to the Command

### PR DESCRIPTION
**Pull request by orta**
Warning funky types ahead. Will comment inline. Fixes #339

### CODE REVIEW

File: src/domain/command.ts [hunk 1]
```diff
@@ -1,19 +1,25 @@
 import { RunContext } from './run-context'
 
-export interface GluegunCommand {
+export interface GluegunCommand<TContext extends RunContext = RunContext> {
```
Comment (orta) [inline src/domain/command.ts]: `TContext` is a generic type, which should extend `RunContext` and if you don't include it the `<XXYY>` after your GluegunCommand, you get `RunContext`

File: src/domain/command.ts [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
+  /** The name of your command */
   name?: string
+  /** A tweet-sized summary of your command */
   description?: string
-  run: (context: RunContext) => void
+  /** The function for running your command, can be async */
+  run: (context: TContext) => void
+  /** Should your command be shown in the listings  */
   hidden?: boolean
   commandPath?: string[]
```
Comment (orta) [inline src/domain/command.ts]: I couldn't find docs internally about this, so someone else will have to doc it

File: src/domain/command.ts (hunk 1 above)
Comment (jamonholmgren) [reply src/domain/command.ts]: This is interesting, I didn't know this was possible in TS. Very cool.

File: src/domain/command.ts (hunk 2 above)
Comment (jamonholmgren) [reply src/domain/command.ts]: I can doc it!

**Pull request merged**